### DISCUSSION TITLE
Switching between Client and Server

### DISCUSSION THREAD

**Discussion by mstegen** [eModbus/eModbus#117] [Q&A] [answered]
First of all thanks for this robust library!

Is there a good way to switch between Client and Server?
My controller has the option to disable Modbus, or to setup a Client/Master or Server/Node (and ID).

When switching from Client to Server (or vice versa), i stop the workers using the following functions:
MBserver.stop(); or
MBclient.end();

Then i setup the new modbus mode, and start the workers again with the corresponding function :
// Start ModbusRTU Master/Client backgroud task
        MBclient.begin();
or 
// Start ModbusRTU Node/Server background task
        MBserver.start();

However, after switching a few times, the ESP32 resets with "CORRUPT HEAP: Bad head at 0x3ffb45e6. Expected 0xabba1234 got 0x12343ffb"

Clearly there must be something else i'm missing here, any idea?

**Comment by Miq1** [comment]
Welcome, and thanks for your warm words 😁 

Yours is an application I never had in mind, so it may well be that there is something fishy in the lib in that regard. I suppose there must be some memory left unreleased that should not be. A couple of allocations are done within the constructors and tasks are started, but with the stop() and end() calls everything is supposed to be cleaned up.

Would you be able to log the free heap whenever you are stopping or starting a client or server? The call is 
```
Serial.println(ESP.getFreeHeap());
```

If the is a memory leak hidden, we should be able to see that.

**Reply by mstegen** [reply]
Thanks for your quick reply.

Although the available heap is slowly decreasing after stopping and starting the client and server, i don't think it's causing the reset.

Could it be that the  `ModbusMessage` data is read in a callback function, while i'm stopping the server, and thus invalidating the data?

**Comment by Miq1** [comment]
In what situation will the reboot happen - stopping the client, stopping the server, restarting the client or restarting the server? May be we can narrow down the problem to one?
Normally all allocated memory should be freed with the task being killed. The callback is run from the client task, so any memory will be given back to the heap.

**Comment by mstegen** [comment, marked as answer]
Ok, i figured it out. It was caused by me calling functions while the server was shut down... That obviously will not work.

I think i found a bug (or feature?) when registering a server handler.

EDIT:

> You may specify different server identifications for registered callbacks. This way the server you are providing can cover more than one Modbus server at a time. You will have to check the requested serverID in the callback to learn which of the registered servers was meant in the request.

OK it's a feature. I did not read the documentation well enough..

**Reply by Miq1** [reply]
Would you need some ``deregisterWorker()`` feature? Should be easy to implement.

**Reply by mstegen** [reply]
Actually, what i really would like is an option to receive all server modbus messages.
Then i can filter/decode it myself.

But even without that, it's working reliable so far. 
Thanks again for this library!

**Reply by Miq1** [reply]
You can use the ANY_FUNCTION_CODE pseudo FC, but you would have to have one ``registerWorker()`` call for each server ID.

I refrained from having a ``ANY_SERVER_ID`` pseudo ID as well, as that would disable a lot of functionality within the lib to catch errors etc.
